CDDA merge (4/x): Connect CD audio codecs into DOSBox and improve how DOSBox uses them

krcroft opened the pull request:
**Runtime improvements:**
- Replaces the existing audio callback routine with an efficient chunked circular-buffer audio reader
- Replaces assumptions that all audio tracks are 44.1 kHz & stereo. The mixer is now fed data at the actual compressed track's rate and channel count
- Eliminates all SDL locks in the audio code in favour of mixer state control
- Queries the codec for track-length instead of using hundreds of iterative decimating seeks to determine track length (loading a 99-track CUE now takes 0.1 user-seconds versus 3+ seconds previously)
- Seeks are performed within the already-decoded buffer (for all codecs) instead of discarding and re-running the decode sequence
- SDL_Sound's buffer-size is now set once and never resized, which eliminates repeated re-malloc'ing in the library
- Only one seek is performed per-playback sequence followed by sequential decodes, similar to a physical CDROM (The baseline dosbox performs a seek for every 2352-bytes of uncompressed audio)
- The DOSBox mixer is now only active during playback sequences and fully dormant otherwise (baseline dosbox instead performs hundreds of calls/second with empty data)
- When using Opus audio tracks, and if your dosbox.conf [mixer] rate=48000, then you will (very likely) achieve sample-exact unadulterated pass-through along your entire audio chain from the decoder, to DOSBox, to your operating system's software mixer, to your sound card driver, to your sound card, to your speakers, or to your digital receiver / USB speakers/headphones / or HDMI TV/screen. This is because almost all modern hardware DACs use a native sample rate of 48000

**Source-level maintenance improvements:**
- Eliminated external dependence on SDL_Sound and its outdated of external library dependencies -- thank you Dominus and DosFreak for recommending this approach.
- SDL_Sound, Vorbis, FLAC, and MP3 libraries are not needed on your system nor do you need to build or provide pointers to these in CFLAGS/LIBS/etc..
- CD image mounting and all codecs are "always on"
- It strips all references to SDL_Sound from configure.ac
- It strips all pre-processor #ifdef branching for SDL_Sound from the code
- It adds src/libs/decoders, which holds the top-level opus, flac, vorbis, and wav decoders
- Fixes all codec compiler warnings; builds have been tested with gcc 4 to 10, Clang 6 to 10, and VisualStudio 14
- Tested on Linux, MacOS X (xcode), and Windows (MSYS2) operating systems
- Tested on i386, x86_64, ARM (little-endian), and PowerPC (big-endian) architectures

krcroft commented:
Oops, I need to add opusfile and speexdsp to the Windows vcpkg lines.

krcroft commented:
Added and repushed; builds underway.
I've smoke-tested the binary on my x86_64 Linux desktop and all is well for all codecs.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
+	return AUDIO_S16MSB;
+#else
+	return AUDIO_S16LSB;
+#endif
```
After looking more in detail in this code… So we have this `getEndian` method, with this implementation ifdefed (the other implementation is not ifdefed, but it's just a getter), and the only usage is also ifdefed, and it's used to overwrite a function pointer indicating which method should be used on a different object…

So this getter is part of overengineered function dispatcher. I don't have a clear idea how to simplify this right now, but let's keep this in mind - there is a room for improvement here.

krcroft replied:
Yes; in `cdda-merge-5`, I've done away with needing any preprocessor statements to deal with endianness, and these endian-related code blocks are now half the lines.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
-		memset(buffer + bytes, 0, count - bytes);
-	} else {
-		memcpy(buffer, sample->buffer, count);
+	return rate;
```
return sample ? sample->actual.rate : 0;

?

krcroft replied:
This is very nice! Fixed in 4e96f038

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
 	}
-	
-	return !(sample->flags & SOUND_SAMPLEFLAG_ERROR);
+	return channels;
```
return sample ? sample->actual.channels : 0;

?

krcroft replied:
Same as above

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
 	
 CDROM_Interface_Image::CDROM_Interface_Image(Bit8u subUnit)
-                      :subUnit(subUnit)
+		      :subUnit(subUnit)
```
Mixed tabs and spaces.

krcroft replied:
nice catch! I need to see if `nano` can highlight prefixed tabs and spaces differently so these things jump out at me!

dreamer replied:
I don't know about nano, but in vim/nvim you can use `:set list` to highlight whitespace characters and `:set listchars=…` to specify what characters should be used for highlighting, I use following setting in my `.vimrc`:

    " Change invisible characters
    set listchars=tab:»\ ,trail:·,eol:↲

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
+		if (player.channel == NULL) {
+			// channel is kept dormant except during cdrom playback periods
+			player.channel = MIXER_AddChannel(&CDAudioCallBack, 0, "CDAUDIO");
+			player.channel->Enable(false);
```
*Idea fo future improvement*: Method `Enable(bool)` is weird, it would be nicer to have methods `Enable()`, `Disable()`.

krcroft replied:
Absolutely. A series of dedicated mass-cleanup patches are probably in order; where each commit or PR changes one specific aspect or style, across the board.

Throw in C++11'isms (say, changing all news & mallocs to unique-pointers, or arrays to vectors, and so on), would be fantastic and likely produce large reductions in LOC and save on complexity.  (side-comment: would be great if we had a cyclomatic complexity analysis check along side the static analysis. Will have to dig around and see what's out there).

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -154,8 +248,10 @@ CDROM_Interface_Image::~CDROM_Interface_Image()
-		player.channel->Enable(false);
+		StopAudio();
+		MIXER_DelChannel(player.channel);
+		player.channel = NULL;
```
`nullptr` ;) ?

krcroft replied:
Fixed in 17c97348, also updated mp3.cpp; the two C++ files part of the patch.

krcroft replied:
This fix rippled a bit :-)

Switching NULL to nullptr produced: _"cdrom_image.cpp:70:14: error: ‘nullptr’ was not declared in this scope"_

Turns out nullptr is only recognized in C++03 and above. Fortunately there's an existing way to  check for it in autotools, which I've added here: https://github.com/dreamer/dosbox-staging/commit/6468970939c4b95bb5eab4ccac006b7aeca8ab24

Tested the change locally and worked great.  However, in CI `./configure` fails.  Turns out that configure.ac statement `AX_CXX_COMPILE_STDCXX_0X` is actually an m4 macro-file that doesn't exist by default.

http://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_cxx_compile_stdcxx_0x.m4

After a bit more digging, I see that the package `autoconf-archive` supplies it. It was also a relief to see that this package is available across the board in all our CI package managers.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -188,6 +289,17 @@ bool CDROM_Interface_Image::GetAudioTracks(int& stTrack, int& end, TMSF& leadOut
+	  leadOut.min,
+	  leadOut.sec,
+	  leadOut.fr);
+#endif
```
Format string is long enough to be split into 2 lines, I think.

Parameters would look better if indented with 2 tabs instead of a tab + spaces.

krcroft replied:
Big cleanup in b6dc0476; everything is aligned properly now regardless of tab width.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -208,13 +332,35 @@ bool CDROM_Interface_Image::GetAudioSub(unsigned char& attr, unsigned char& trac
+	  relPos.min,
+	  relPos.sec,
+	  relPos.fr);
+#endif
```
Mix of tabs and spaces in formatting.

krcroft replied:
Got it

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -223,39 +369,130 @@ bool CDROM_Interface_Image::GetMediaTrayStatus(bool& mediaPresent, bool& mediaCh
+		LOG(LOG_MISC, LOG_WARN)("Game tried to read sector %lu, which is beyond the 100-minute maximum of a CDROM", start);
+
+	// We can't play audio from a data track (as it would result in garbage/static)
+	else if(track >= 0 && tracks[track].attr == 0x40)
```
Is there some enum indicating 0x40 to be a data track?

krcroft replied:
0x40 identifies a data track (0x41 is a lead-in track, and others). There unfortunately isn't an enum set locking down these magic identifiers, but there should be (DOSBox is riddled with unexplained, unreferenced magic numbers just like this).

Unfortunately I'm similarly worried about "change overreach" if I start changing more files (cdrom.cpp, cdrom_aspi_win32.cpp, and cdrom_image.cpp) than are needed to impart the feature's I'm added.

Sigh.  So much fear and uncertainty because of essentially little or no communication.  Hopefully this gets cleared up soon so the pathway is clear - one way or another.

dreamer replied:
Understood. I agree - let's get this merged before branching out to improvements in other areas.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -223,39 +369,130 @@ bool CDROM_Interface_Image::GetMediaTrayStatus(bool& mediaPresent, bool& mediaCh
+#else
+			if (trackFile->getEndian() == AUDIO_S16SYS)
+#endif
+				player.addSamples = channels ==  2  ? &MixerChannel::AddSamples_s16 \
```
Callback to my earlier comment about getEndian() ^^^

krcroft replied:
The preprocessors are now gone in cdda-merge-5; although I'm still using getEndian to assess each track.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -223,39 +369,130 @@ bool CDROM_Interface_Image::GetMediaTrayStatus(bool& mediaPresent, bool& mediaCh
+			   len,
+			   player.playbackRemaining / (1000 * bytesPerMs)
+			);
+#endif
```
Mix of tabs and spaces here.

krcroft replied:
Fixed in prior commit.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -313,58 +549,140 @@ bool CDROM_Interface_Image::ReadSector(Bit8u *buffer, bool raw, unsigned long se
+	LOG_MSG("\r%3d%% [%.*s%*s] - %s", val, lpad,
+	   "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||", rpad, "", msg);
+	fflush(stdout);
+}
```
Is this function still useful?

krcroft replied:
Knocked it out along with the commented-out calls to it; it's past its usefulness.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -530,26 +852,16 @@ bool CDROM_Interface_Image::LoadCueSheet(char *cuefile)
-				success = false;
-			}
+            else
+                track.file = new AudioFile(filename.c_str(), error);
```
I think you noticed already, but Clang static analyzer indicates a memory leak in here (the interaction triggering memory leak is very complicated and requires at least 3 loop iterations, IIRC. Anyways, I think it can be easily resolved by using C++11 `std::unique_ptr`.

krcroft replied:
If `track.file` was created in the local-scope, then this would be a textbook use, but `track.file` is a member variable (and a common one for two derived classes), so we'd have to change its type to `unique_ptr` up in the class definition.  I'm a bit reluctant only because I'm worried it will be another "mark against" the code for merging upstream.. but depending on what acceptance we see, and if upstream is willing to accept these C++ "upgrades", then I will go through this Audio code and move all of the member pointers to smart pointers like this (and do-away with delete across the board).  We'll see!

krcroft replied:
"won't fix" for now, but tentative for future changes that "upgrade" to newer C++'isms.

krcroft replied:
PS- these two "fromISO" functions are a mess of spaghetti.. looking forward to re-writing them too.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -558,33 +870,38 @@ bool CDROM_Interface_Image::LoadCueSheet(char *cuefile)
 	track.length = 0;
 	track.file = NULL;
-	if(!AddTrack(track, shift, 0, totalPregap, 0)) return false;
+	if(!AddTrack(track, shift, -1, totalPregap, 0)) return false;
```
While touching this line, let's fix the formatting ;) (space between `if` and `(`). Also, return will look better on a new line.

krcroft replied:
Fixed all of these in cdrom_image, and brought all the dangling returns under the conditional.
Doesn't look as elegant in some places, but it's at least consistent and some of those returns are far more obvious than being dangling out in the far-right of the text.
Fixed in 859cee72

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -558,33 +870,38 @@ bool CDROM_Interface_Image::LoadCueSheet(char *cuefile)
+	if (prestart >= 0) {
 		if (prestart > curr.start) return false;
 		skip = curr.start - prestart;
 	} else skip = 0;
```
`if` got touched, then let's fix formatting in line 903:

    } else {
        skip = 0;
    }

Alternatively, `skip` can be initialized to 0, making `else` unnecessary.

krcroft replied:
Fixed in bbc159a9.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -594,34 +911,44 @@ bool CDROM_Interface_Image::AddTrack(Track &curr, int &shift, int prestart, int
+	LOG_MSG("%s CDROM: AddTrack cur.start=%d cur.len=%d cur.start+len=%d | prev.start=%d prev.len=%d prev.start+len=%d",
+	        get_time(),
+	        curr.start, curr.length, curr.start + curr.length,
+	        prev.start, prev.length, prev.start + prev.length);
```
Tabs got mixed with spaces (but in this case it actually looks ok). Up to you if you want to replace it with tabs, if this wasn't intentional.

krcroft replied:
Yes, this is intentional and how I've tried to line up sub-lines.
```
<tab><tab><tab>First Line Pre-words (x<- position to line-up sub-lines
<tab><tab><tab><..... spaces .....  >Second line
<tab><tab><tab><..... spaces .....  >Third line );
```
Using spaces after the same number of prior-tabs keeps the text alignment regardless of tab-width.  I've made all the LOG_MSG's consistent with this now.

dreamer commented on src/libs/decoders/stb_vorbis.h:
```diff
@@ -78,6 +78,12 @@
+extern "C" {
+#endif
+*/
+
```
To be removed, I guess?

krcroft replied:
Yes!

dreamer commented on src/libs/decoders/stb_vorbis.h:
```diff
@@ -1301,6 +1339,15 @@ static uint8 get8(vorb *z)
+   if (SDL_RWread(z->rwops, &c, 1, 1) != 1) { z->eof = TRUE; return 0; }
+   return c;
+   }
+   #endif
```
Formatting of this block is a bit inconsistent.

krcroft replied:
Great catch.  Fixed in 003a3fd2.
Yes (also interesting; looks like STB uses indent sizes of three-spaces.)

dreamer commented on src/libs/decoders/stb_vorbis.h:
```diff
@@ -1329,6 +1376,14 @@ static int getn(vorb *z, uint8 *data, int n)
       return 1;
    }
 
+   #ifdef __SDL_SOUND_INTERNAL__
```
ifdefs to first column?

krcroft replied:
STB indents preprocessors with the code level (atleast for the majority, he does have a few inconsistencies).  In this case, I want to match STB's style instead of DOSBox's.  If we were forking the STB files and taking them in our own direction, then we could make them 100% DOSBox' compliant. (but instead, we'll always be taking new STB files from upstream, and peppering these SDL changes into them, so prefer to follow his style).  Unfortunately with STB, there is no hope of up-streaming our changes, so we're stuck in this middle-ground.

dreamer commented on src/libs/decoders/stb_vorbis.h:
```diff
@@ -3834,8 +3913,8 @@ static int start_decoder(vorb *f)
                   int off = (z / div) % c->lookup_values;
-                  float val = mults[off];
-                  val = mults[off]*c->delta_value + c->minimum_value + last;
+                  // float val = mults[off]; // under review at https://github.com/nothings/stb/issues/816
```
Thanks for this link… I am a bit worried by the number of open issues and unmerged pull requests in STB's repo :(

krcroft replied:
Although it looks unkept on the surface, his repo gets a ton of traffic and he does make his way through the reports, albeit slower than Dave of dr_libs.

For us though  stb_vorbis is an early creation from 12 years ago (pre-Opus) and was funded by RAD game tools (now used in over 23,000 games!).  So the library is extremely battle-hardened and functional fixes are quite rare now.  Sean does address them; two versions ago I report the skeleton ogg issue (which became v1.15 of stb_vorbis) and a patch, which he accepted:

https://github.com/nothings/stb/commit/604b9367ee7a77aeff916a6bb5a9ee9601d72c02

Dr_libs are much younger and I suspect used more rarely (which is unfortunate, because they are of impeccable quality and performance) For example, dr_mp3 hadn't been tested (and didn't work) on my big-endian PPC7400, and Dave was very interested in getting that working. Same with the fast seeking; that was some awesome he did and we collaborated last winter.

I'm rambling now, but yes, when Sean leaves issues open, it's my interpretation that those are the ones he plans to address in the next revision cycle. Might be another 6 months or so before that happen with stb_vorbis though!

krcroft replied:
Another very exciting this about stb's libraries is that they have been selected by Google's fuzzing team.  I would love to see all the codecs fuzzed. So far Sean hasn't responded Google's offer, but he also hasn't closed it, whcih tells me he's probably interested (maybe if someone submit a patch). 
This criteria is pretty significant too:
"In order for a project to be accepted to OSS-Fuzz, it needs to have a large user base and/or be critical to Global IT infrastructure"

dreamer commented on visualc_net/dosbox.vcxproj:
```diff
@@ -69,7 +69,7 @@
     </ClCompile>
     <Link>
-      <AdditionalDependencies>opengl32.lib;sdl_net.lib;winmm.lib;libpng16d.lib;sdlmaind.lib;sdl.lib;odbc32.lib;odbccp32.lib;ws2_32.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>opengl32.lib;sdl_net.lib;winmm.lib;libpng16d.lib;sdlmaind.lib;sdl.lib;odbc32.lib;odbccp32.lib;ws2_32.lib;opusfile.lib;opus.lib;ogg.lib;libspeexdsp.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
I suggest putting changes to VisualStudio files in a sperate commit - it will make it easier to submit the changes upstream.

krcroft replied:
Separated into 6b6344b3.

dreamer commented:
Have you ever tested new implementation with a game that requires changing discs mid-game? I know only 1 such game, but it has discs full of FMVs and not audio. I saw nothing indicating, that it would be broken in such case - I am just curious.

krcroft commented:
Documentation updates and some more comment fixes made in 8f2e1d60.

krcroft commented on src/dos/cdrom.h:
```diff
@@ -154,37 +155,51 @@ class CDROM_Interface_Image : public CDROM_Interface
+		virtual Bit32u getRate() = 0;
+		virtual Bit8u  getChannels() = 0;
+		virtual int    getLength() = 0;
+		virtual        ~TrackFile() { };
```
Excellent; I prefer `default`over `{ }` because the later might imply that nothing is destructed as opposed to tearing down member objects and baseclasses (which `default` does). 
Actually, would it make more sense for me to delete this destructor all together - which implies default?

krcroft replied:
Fixed in 66fdb679

krcroft commented on src/dos/cdrom_image.cpp:
```diff
@@ -16,6 +16,11 @@
 
+// #define DEBUG 1
+#ifdef DEBUG
+#include <time.h>
```
I wound up knocking out the std::chrono timer from seek's DEBUG output; so knocked out <ctime> now that I'm only using the old-school time_t and tm structs.

krcroft commented on src/dos/cdrom_image.cpp:
```diff
@@ -43,7 +48,21 @@ using namespace std;
+
+	time(&rawtime);
+	ptime = localtime(&rawtime);
+	snprintf(time_str, 9, "%02d:%02d:%02d", ptime->tm_hour, ptime->tm_min, ptime->tm_sec);
```
(argh; I'm not sure if github just collapses these reviews away - even when I click the "14 hidden" link, but somehow I miss some and they only appear when I re-open the page. )
Fixed in fdd331bb.

krcroft commented:
> Have you ever tested new implementation with a game that requires changing discs mid-game? I know only 1 such game, but it has discs full of FMVs and not audio. I saw nothing indicating, that it would be broken in such case - I am just curious.

Tested it in my EXO archive. _Wing Commander IV_ has six CDs (wow; DVD didn't arrive soon enough for them!).  
```
imgmount d -t iso cd1.iso cd2.iso cd3.iso
d:
dir
<ctrl + F4>
dir
<ctrl + F4>
dir
<ctrl + F4>
dir
<ctrl + F4>
```
You will also need DOSBox console output mentioning:
```
Drive D: disk 2 of 3 now active
Diskcaching reset for normal mounted drives.
Drive D: disk 3 of 3 now active
Diskcaching reset for normal mounted drives.
Drive D: disk 1 of 3 now active
Diskcaching reset for normal mounted drives.
Drive D: disk 2 of 3 now active
```

krcroft commented:
The fixes grew the scope of this PR, but we're in much better shape now with C++03 enabled more fixes put into configure.ac and the source to support it.

dreamer commented on INSTALL:
```diff
@@ -32,12 +50,6 @@ SDL_Net (optional)
-
 ALSA_Headers
     (optional)
     for Alsa support under linux. Part of the linux kernel sources
```
Turns out GitHub interface does not allow reviewer to attach comments to lines outside of diff context…

About lines 58, 59:

    If you want compile from developer sources (SVN) under a unix system, you'll also need 
    automake (>=1.6), autoconf(>=2.50). Should be available at http://www.gnu.org

Mention `autoconf-archive` here, please. I did it originally in a patch, that was sent to SVN upstream in previous series (5722a4e4fe), but this change is "in limbo" - neither accepted by Qbix nor refused.

krcroft replied:
Fixed in 9bff690fd07b4bd975289795a850a64af4979fda

dreamer commented on src/dos/cdrom.h:
```diff
@@ -247,6 +267,7 @@ static  struct imagePlayer {
 	bool	GetCueKeyword(std::string &keyword, std::istream &in);
 	bool	GetCueFrame(int &frames, std::istream &in);
 	bool	GetCueString(std::string &str, std::istream &in);
+	// bool	AddTrack(Track &curr, int &shift, int prestart, int &totalPregap, int currPregap, int frameFromCue);
```
Probably not needed.

krcroft replied:
Fixed in a3dc6a407ca75a89293070e24823e33ab1d783c6.

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -43,16 +47,30 @@ using namespace std;
 {
 	file = new ifstream(filename, ios::in | ios::binary);
-	error = (file == NULL) || (file->fail());
+	error = (file == nullptr) || (file->fail());
```
*Just a comment, there's no need to change the code, I hope.*

Previously, I asked for usage of nullptr here, because I thought we might go for C++11 again. I am a bit surprised nullptr compiles with C++03 - the feature was added in C++11, but maybe nullptr itself existed already and only `nullptr_t` got added in new standard? Anyways, compilers do not seem to object to this, so maybe it's fine…

krcroft replied:
From some reading, C++03 or c++0x was a long-time evolving standard the compilers tried to keep up with. You can see GCC's support for it evolve from early 4.x versions through to 4.8.1, here: https://gcc.gnu.org/projects/cxx-status.html#cxx11

When C++03 was finalized it was renamed C++11, and the C++03 standard was dropped/deprecated.  It just happens that if you use a modern-enough compiler that fully supported the final version of C++03 (such as gcc 4.8.1) then you wind up with full C++11 support as well :-).  Compilers newer than 4.8.1 and Clang 3.3 essentially given you full C++11 support when you ask for the c++0x standard.

If our CI workflow hit pre-4.8.1 compilers, then we might find one that doesn't support nullptr(_t).  Clang seems even more advanced.. we'd have to go back to pre-3.3 to exclude C++11 support!

dreamer replied:
Ah, sorry, I should've read your changes more thoroughly and respond here before merging in - I have no problem with using C++11, so it gave me even less incentive to double-check what happened. And I forgot that C++0x brouhaha happened 10 years ago already and new generation of programmers might not know about it.

C++0x is not C++03! It is the next standard AFTER C++03. It was called 0x, because work on it started ~2005-2006 and committee expected it to be released in a year or two. 2008 came and went, then 2009 came and went, 2010 came and went and the standard was finally released in 2011. The compiler flag is preserved for backwards-compatibility. C++0x is "beta" version of C++11.

krcroft replied:
Maybe we should flip over that configure.ac macro to just give us C++11? 

Maybe Qbix isn't aware of this detail when he mentioned 0x would be fine? (perhaps he thought it was to some prior early-2000's version?) Otherwise if he really is fine with beta-C++11 then why not just go with the finalized version..

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -305,110 +573,205 @@ int CDROM_Interface_Image::GetTrack(int sector)
+		//	     we need to wrap any remaining decoded-but-not-consumed
+		//	     samples back around to the front of the buffer.
+		//
+		// 3. Fill: When out circular buffer is too depleted to satisfy the
```
Typo? `out` was probably `our`?

krcroft replied:
559de7d454c79ea6d88420d000a51e170d1ac4daFixed in

krcroft commented:
Will now squash into a group of distinct commits...

krcroft commented:
Squashed and split the commits into groups of files:
- 6d8c9b93: dosbox-staging's CI and scripts (internal only)
- b932a89f: upstream library customization (stb_vorbis)
- 29bcde1d: minor updates to our decoders
- 7ee4a20c: changes to DOSBox's existing CDROM-handling sources
- 90238d2f: additions to DOSBox's existing autotools configuration and documentation
- fbfd5632: additions to DOSBox's existing VisualStudio project file

dreamer commented:
This is now merged in master, as of 7f3b14c99fffdf4d6bf2b3c18b27e84d26037a72.

GitHub haven't noticed, that code got merged, so closing the review manually.

krcroft commented:
Thanks for the kind words and support @dreamer ; knowing that this is helping beyond it's niche focus really makes it worthwhile!
